expression/builtin_grouping.go

```diff
@@ -0,0 +1,191 @@
+// limitations under the License.
+
+package expression
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+var (
+	_ functionClass = &groupingFunctionClass{}
+)
+
+var (
+	_ builtinFunc = &builtinGroupingSig{}
+)
+
+type groupingFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *groupingFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	argTp := []types.EvalType{types.ETInt}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, argTp...)
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(1)
+	sig := &builtinGroupingSig{bf, 0, map[int64]int64{}}
+	sig.setPbCode(tipb.ScalarFuncSig_GroupingSig)
+	return sig, nil
+}
+
+type builtinGroupingSig struct {
+	baseBuiltinFunc
+
+	// TODO these are two temporary fields for tests
+	version     uint32
+	groupingIDs map[int64]int64
+}
+
+func (b *builtinGroupingSig) SetMetaVersion(version uint32) {
+	b.version = version
+}
+
+func (b *builtinGroupingSig) SetMetaGroupingIDs(groupingIDs map[int64]int64) {
+	b.groupingIDs = groupingIDs
+}
+
+func (b *builtinGroupingSig) getMetaVersion() uint32 {
+	return b.version
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingSig) metadata() proto.Message {
+	args := &tipb.GroupingFunctionMetadata{
+		// TODO
+	}
+	return args
+}
+
+func (b *builtinGroupingSig) Clone() builtinFunc {
+	newSig := &builtinGroupingSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.version = b.version
+	newSig.groupingIDs = b.groupingIDs
+	return newSig
+}
+
+func (b *builtinGroupingSig) getMetaGroupingIDs() map[int64]int64 {
+	return b.groupingIDs
+}
+
+func (b *builtinGroupingSig) getMetaGroupingID() int64 {
+	var metaGroupingID int64
+	grouping_ids := b.getMetaGroupingIDs()
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingSig) checkMetadata() error {
+	version := b.getMetaVersion()
+	grouping_ids := b.getMetaGroupingIDs()
+	if version < 1 || version > 3 {
+		return errors.Errorf("Version of meta data in grouping function is invalid. input version: %d", version)
+	} else if (version == 1 || version == 2) && len(grouping_ids) != 1 {
+		return errors.Errorf("Invalid number of groupingID. version: %d, number of groupingID: %d", version, len(b.groupingIDs))
+	}
+	return nil
+}
+
+func (b *builtinGroupingSig) groupingImplV1(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID&metaGroupingID > 0 {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingSig) groupingImplV2(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID > metaGroupingID {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingSig) groupingImplV3(groupingID int64) int64 {
```

Comment:
```
| year | country | product    | profit | grp_year | grp_country | grp_product |
| 2001 | NULL    | NULL       |   3010 |        0 |           1 |           1 |
```
for a case like the above, the grouping(product) = 1 means the product col is filled as null for a bigger group granularity of `a`. And the actual group column is a, so we pass the corresponding grouping_id as parameter as a meta (form a id list here) to grouping function here. Make sense.

for V2, it seems rarely used ( control the grouping id > some value), it's ok, not a strong requirement